copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+					}
+					tw.Flush()
+				}
+				hdr, err = tr.Next()
+			}
+			if err != io.EOF {
+				return errors.Wrapf(err, "error extracting contents of archive %s", contentPath)
+			}
+			return nil
+		}
+		// if this regular file is hard linked to something else we've
+		// already added, set up to output a TypeLink entry instead of
+		// a TypeReg entry
+		target := hardlinkChecker.Check(srcfi)
+		if target != "" {
+			hdr.Typeflag = tar.TypeLink
+			if filepath.Dir(filepath.Clean(string(os.PathSeparator)+name)) == filepath.Dir(target) {
+				target = filepath.Base(target)
+			}
+			hdr.Linkname = target
+			hdr.Size = 0
+		} else {
+			// note the device/inode pair for this file
+			hardlinkChecker.Add(srcfi, string(os.PathSeparator)+name)
+		}
+	}
+	// map the ownership for the archive
+	if idMappings != nil && !idMappings.Empty() {
+		hostPair := idtools.IDPair{UID: hdr.Uid, GID: hdr.Gid}
+		hdr.Uid, hdr.Gid, err = idMappings.ToContainer(hostPair)
+		if err != nil {
+			return errors.Wrapf(err, "error mapping host filesystem owners %#v to container filesystem owners", hostPair)
+		}
+	}
+	// output the header
+	if err = tw.WriteHeader(hdr); err != nil {
+		return errors.Wrapf(err, "error writing header for %s (%s)", contentPath, hdr.Name)
+	}
+	if hdr.Typeflag == tar.TypeReg {
+		// output the content
+		f, err := os.Open(contentPath)
+		if err != nil {
+			return errors.Wrapf(err, "error opening %s", contentPath)
+		}
+		defer f.Close()
+		n, err := io.Copy(tw, f)
+		if err != nil {
+			return errors.Wrapf(err, "error copying %s", contentPath)
+		}
+		if n != hdr.Size {
+			return errors.Errorf("error copying %s: incorrect size (expected %d bytes, read %d bytes)", contentPath, n, hdr.Size)
+		}
+		tw.Flush()
+	}
+	return nil
+}
+
+func copierHandlerPut(bulkReader io.Reader, request Request, idMappings *idtools.IDMappings) (*Response, func() error, error) {
+	errorResponse := func(fmtspec string, args ...interface{}) (*Response, func() error, error) {
+		return &Response{Put: PutResponse{Error: fmt.Sprintf(fmtspec, args...)}}, nil, nil
+	}
+	dirUID, dirGID := 0, 0
+	if request.PutOptions.ChownDirs != nil {
+		dirUID, dirGID = request.PutOptions.ChownDirs.UID, request.PutOptions.ChownDirs.GID
+	}
+	dirMode := os.FileMode(0755)
+	if request.PutOptions.ChmodDirs != nil {
+		dirMode = *request.PutOptions.ChmodDirs
+	}
+	var fileUID, fileGID *int
+	if request.PutOptions.ChownFiles != nil {
+		fileUID, fileGID = &request.PutOptions.ChownFiles.UID, &request.PutOptions.ChownFiles.GID
+	}
+	if idMappings != nil && !idMappings.Empty() {
+		containerDirPair := idtools.IDPair{UID: dirUID, GID: dirGID}
+		hostDirPair, err := idMappings.ToHost(containerDirPair)
+		if err != nil {
+			return errorResponse("copier: put: error mapping container filesystem owner %d:%d to host filesystem owners: %v", dirUID, dirGID, err)
+		}
+		dirUID, dirGID = hostDirPair.UID, hostDirPair.GID
+		if request.PutOptions.ChownFiles != nil {
```

Comment:
The API documentation does not suggest any asymmetry between `ChownDirs` and `ChownFiles`; why does the code does not look exactly the same?

Reply:
We need defaults for directory ownership for the sake of any intermediate directories that we need to create that aren't part of the archive that we're extracting, but we want to use the supplied values if any are specified for directories.